lib/navigation/preprocessing.cpp

```diff
@@ -1,123 +1,129 @@
-    /*
-    TODO: implement
-    rough process:
-    - set as class member (?)
-    - check if keyence sensors agree
-    - if they agree, set keyence disgreeemnt to false
-    - if they disagree and keyenece_disagreement = false, set to true
-    - if they disagree and keyenece_disagreement = true, 
-    we have 2 disgeements in a row so fail state
-    - update with largest of sensor values
-    
-    -also figure out return type/ what we update and update
-    documentation as appropriate
-    */
-   return {0, 0};
-  }
-
-  core::ImuData Preprocessing::preprocessImus(const core::RawImuData imu_data){
-
-    /*
-    TODO: implement
-    rough plan:
-    - set raw data as class member
-    - assign imu_data to class object
-    - run imu outlier detection
-    - run check imus reliable
-    - get mean and run kalman filter
-    - update trajectory with mean
-
-    -also figure out return type/ what we update and update
-    documentation as appropriate
-    */
-   return {0.0, 0.0, 0.0, 0.0};
-  }
-
-  core::EncoderData Preprocessing::preprocessEncoders(const core::EncoderData encoder_data){
-
-    /*
-    TODO: implement
-    rough plan:
-
-    - assign encoder data to class object
-    - run encoder outlier detection
-    - run check encoders reliable
-    - update trajectory with mean
-
-    -also figure out return type/ what we update and update
-    documentation as appropriate
-    */
-   return {0, 0, 0, 0};
-  }
-}
\ No newline at end of file
+namespace hyped::navigation {
+
+Preprocessing::Preprocessing()
+{
+  // TODO: implement
+}
+
+void Preprocessing::imuOutlierDetection()
+{
+  /*
+  TODO: implement
+  rough process:
+  - get q1, median, q3 of imu array
+  - define upper & lower bounds as (-)1.5*inter-quatrile range
+  - if any datapoint outwith this, set
+  outlier_imus_[i] += 1
+  -set outlier points as median
+
+  -also has to be able to handle 1 unreliable sensor
+  -also figure out return type/ what we update and update
+  documentation as appropriate
+  */
+}
+
+void Preprocessing::encoderOutlierDetection()
+{
+  /*
+  TODO: implement
+  rough process:
+  - get q1, median, q3 of encoder array
+  - define upper & lower bounds as (-)1.5*inter-quatrile range
+  - if any datapoint outwith this, set
+  outlier_imus_[i] += 1
+  -set outlier points as median
+
+  -also has to be able to handle 1 unreliable sensor
+  -also figure out return type/ what we update and update
+  documentation as appropriate
+  */
+}
+
+void Preprocessing::checkImusReliable()
+{
+  /*
+  TODO: implement
+  rough process:
+  - check how many times an individual imu
+  has been an outlier in a row (outlier imus)
+  -if outlier_imus_[i] > n (tbd), mark imu as unreliable
+  in reliable imus.
+  -if number unreliable in reliable_imus > 1, fail state
+
+  -also figure out return type/ what we update and update
+  documentation as appropriate
+  */
+}
+
+void Preprocessing::checkEncodersReliable()
```

Comment:
You're working with state too much. This should be a function taken some raw encoder data and returning whether they the data points agree or not. Possibly with past data? (E.g. encoders should never decrease so that's something that would rely on past state)

However, what is the caller going to do with this information? Just knowing that the encoders are faulty doesn't tell me how to fix it. Are we going to throw away the whole reading? Or are we going to compensate somehow?

I would argue that the transformation in `preprocessEncoders` should go from `std::array<uint32_t, kNumEncoders>` to `std::vector<uint32_t>` where we discard all the values that we deem safe (possibly none!). I don't think late stages will care where each reading came from, only how many we have.

Reply:
This checks if an individual encoder is consistently faulty. If so, there  is  no fixing, we just disregard any further readings from that sensor. I see what you mean about probably working with state too much but the idea of working with state here is to keep track of which sensors are giving bad data. I'll modify the new classes I'm going to make to be less reliant on states.

Reply:
Just changed implementation. No raw data is at any point stored in preprocessing classes. Data is passed in, processed and output. The only states used are tracking credibility of sensors